microcosm_pubsub/matchers.py

```diff
@@ -0,0 +1,95 @@
+from dataclasses import dataclass
+from json import loads
+from typing import List, Optional, Mapping
+
+from hamcrest import not_none
+from hamcrest.library.collection.issequence_containinginorder import IsSequenceContainingInOrder
+from hamcrest.library.object.hasproperty import IsObjectWithProperty
+from hamcrest.core.helpers.wrap_matcher import wrap_matcher
+
+
+@dataclass
+class PublishedMessage:
+    """
+    A published message.
+
+    """
+    topic: str
+    message: Mapping[str, str]
+
+    @property
+    def media_type(self) -> Optional[str]:
+        return self.message.get("mediaType")
+
+    @property
+    def uri(self) -> Optional[str]:
+        return self.message.get("uri")
+
+    @classmethod
+    def from_call(cls, call) -> "PublishedMessage":
+        name, args, kwargs = call
+        topic = kwargs["TopicArn"]
+        message = loads(kwargs["Message"])
+        return cls(
+            topic=topic,
+            message=message,
+        )
+
+    @staticmethod
+    def from_mock_calls(sns_producer) -> List["PublishedMessage"]:
+        return [
+            PublishedMessage.from_call(mock_call)
+            for mock_call in sns_producer.sns_client.publish.mock_calls
+            # NB: _Call is a tuple of (name, args, kwargs)
+            if "TopicArn" in mock_call[-1] and "Message" in mock_call[-1]
+        ]
+
+
+class PublishedMessageMatcher(IsObjectWithProperty):
+    """
+    Matcher that validates a single published message.
+
+    """
+    def describe_to(self, description):
+        return description.append_text(
+            f"a message with a {self.property_name} matching ",
+        ).append_description_of(
+            self.value_matcher,
+        )
+
+
+class PublishingMatcher(IsSequenceContainingInOrder):
```

Comment:
The main value we're adding is L68: delegate to the superclass (which is the implementation of `contains`), but translate the (mocked) SNSProducer to a a list of `PublishedMessage` object.